File: Day8/merge2sortedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(l1 == NULL && l2 == NULL){return NULL;}
        else if(l1==NULL && l2){return l2;}
        else if(l1 && l2 == NULL){return l1;}
        
        ListNode* head = new ListNode();
        ListNode* ptr = head;
        
        while(l1 && l2){
            
            if(l1-> val < l2-> val){
                ptr-> next = l1;
                l1 = l1-> next;
            }
            else{
                ptr-> next = l2;
                l2 = l2-> next;
            }
            ptr = ptr-> next;
            
        }
        
        if(l1)
            ptr-> next = l1;
        if(l2)
            ptr->next = l2;
        
        ptr = head-> next;
        delete(head);
        return ptr;
    }
};

File: Day8/merge2sortedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null && l2 == null) {return null;}                
        if (l1 == null)  {return l2;}         
        if (l2 == null) {return l1;}
        
        ListNode head = new ListNode(0, null);
        ListNode ptr;
        ptr = head;
        
        while(l1 != null && l2 != null){
            
            if(l1.val <= l2.val){
                ptr.next = l1;
                l1 = l1.next;
            }
            else{
                ptr.next = l2;
                l2 = l2.next;
            }
            ptr = ptr.next;
        }
        
        if(l1 == null && l2 != null){
            ptr.next = l2;
        }
        else if(l1!= null && l2 == null){
            ptr.next = l1;
        }
        
        head = head.next;
        return head;
    }
}
